sdk/identity/azure-identity/src/token_credential_impl.cpp

```diff
@@ -152,44 +154,96 @@ AccessToken TokenCredentialImpl::ParseToken(
     std::string const& expiresInPropertyName,
     std::string const& expiresOnPropertyName)
 {
-  try
+  auto const parsedJson = Azure::Core::Json::_internal::json::parse(jsonString);
+
+  if (!parsedJson.contains(accessTokenPropertyName)
+      || !parsedJson[accessTokenPropertyName].is_string())
   {
-    auto const parsedJson = Azure::Core::Json::_internal::json::parse(jsonString);
+    ThrowJsonPropertyError(accessTokenPropertyName);
+  }
 
-    if (!parsedJson.contains(accessTokenPropertyName))
-    {
-      ThrowMissingJsonPropertyError(accessTokenPropertyName);
-    }
+  AccessToken accessToken;
+  accessToken.Token = parsedJson[accessTokenPropertyName].get<std::string>();
+  accessToken.ExpiresOn = std::chrono::system_clock::now();
 
-    AccessToken accessToken;
-    accessToken.Token = parsedJson[accessTokenPropertyName].get<std::string>();
+  if (parsedJson.contains(expiresInPropertyName))
+  {
+    auto const& expiresIn = parsedJson[expiresInPropertyName];
 
-    if (parsedJson.contains(expiresInPropertyName))
+    if (expiresIn.is_number_unsigned())
     {
+      // 'expires_in' as number (seconds until expiration)
       accessToken.ExpiresOn
-          = std::chrono::system_clock::now()
-          + std::chrono::seconds(
-                parsedJson[expiresInPropertyName].get<std::chrono::seconds::duration::rep>());
+          += std::chrono::seconds(expiresIn.get<std::chrono::seconds::duration::rep>());
+
+      return accessToken;
     }
-    else if (expiresOnPropertyName.empty())
+
+    if (expiresIn.is_string())
     {
-      ThrowMissingJsonPropertyError(expiresInPropertyName);
+      try
+      {
+        // 'expires_in' as numeric string (seconds until expiration)
+        accessToken.ExpiresOn += std::chrono::seconds(std::stoi(expiresIn.get<std::string>()));
+
+        return accessToken;
+      }
+      catch (std::exception const&)
```

Comment:
Under what circumstances would `stol` throwing be allowed? If there's a string value of expiresIn which *isn't* a number, shouldn't we be rejecting the token?

Reply:
There still can be a readable value in `expires_on`, so we don't throw just yet. The scenario being - some server starts returning floats in `expires_in`.